// ng-zorro-antd/esm5/list/list-cell.js
/**
 * @fileoverview added by tsickle
 * Generated from: list-cell.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, Directive, Input } from '@angular/core';
var NzListEmptyComponent = /** @class */ (function () {
    function NzListEmptyComponent() {
    }
    NzListEmptyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-list-empty',
                    exportAs: 'nzListHeader',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <nz-embed-empty [nzComponentName]=\"'list'\" [specificContent]=\"nzNoResult\"></nz-embed-empty>\n  ",
                    host: {
                        class: 'ant-list-empty-text'
                    }
                }] }
    ];
    NzListEmptyComponent.propDecorators = {
        nzNoResult: [{ type: Input }]
    };
    return NzListEmptyComponent;
}());
export { NzListEmptyComponent };
if (false) {
    /** @type {?} */
    NzListEmptyComponent.prototype.nzNoResult;
}
var NzListHeaderComponent = /** @class */ (function () {
    function NzListHeaderComponent() {
    }
    NzListHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-list-header',
                    exportAs: 'nzListHeader',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <ng-content></ng-content>\n  ",
                    host: {
                        class: 'ant-list-header'
                    }
                }] }
    ];
    return NzListHeaderComponent;
}());
export { NzListHeaderComponent };
var NzListFooterComponent = /** @class */ (function () {
    function NzListFooterComponent() {
    }
    NzListFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-list-footer',
                    exportAs: 'nzListFooter',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <ng-content></ng-content>\n  ",
                    host: {
                        class: 'ant-list-footer'
                    }
                }] }
    ];
    return NzListFooterComponent;
}());
export { NzListFooterComponent };
var NzListPaginationComponent = /** @class */ (function () {
    function NzListPaginationComponent() {
    }
    NzListPaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-list-pagination',
                    exportAs: 'nzListPagination',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <ng-content></ng-content>\n  ",
                    host: {
                        class: 'ant-list-pagination'
                    }
                }] }
    ];
    return NzListPaginationComponent;
}());
export { NzListPaginationComponent };
var NzListLoadMoreDirective = /** @class */ (function () {
    function NzListLoadMoreDirective() {
    }
    NzListLoadMoreDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'nz-list-load-more',
                    exportAs: 'nzListLoadMoreDirective'
                },] }
    ];
    return NzListLoadMoreDirective;
}());
export { NzListLoadMoreDirective };
var NzListGridDirective = /** @class */ (function () {
    function NzListGridDirective() {
    }
    NzListGridDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'nz-list[nzGrid]',
                    host: {
                        class: 'ant-list-grid'
                    }
                },] }
    ];
    return NzListGridDirective;
}());
export { NzListGridDirective };
